File: LeetCode/DataStruct/Golang/Leetcode901-online-stock-span.go
package sbgo

type StockSpanner struct {
  prices, day []int
  idx int 
}


func Constructor() StockSpanner {
  return StockSpanner{
    prices:[]int{int(1e6)},
    day: []int{-1},
    idx: 0,
  }
}


func (this *StockSpanner) Next(price int) int {
  for price >= this.prices[len(this.prices) - 1] {
    this.prices = this.prices[:len(this.prices) - 1]
    this.day = this.day[:len(this.day) - 1]
  }
  res := this.idx - this.day[len(this.day) - 1]
  this.day = append(this.day, this.idx)
  this.prices = append(this.prices, price)
  this.idx++
  return res
}

File: LeetCode/DataStruct/Rust/Leetcode901-online-stock-span.rs
struct StockSpanner {
    prices: Vec<i32>,
    day: Vec<i32>,
    idx: i32,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl StockSpanner {

    fn new() -> Self {
        StockSpanner{
            prices: vec![i32::MAX],
            idx: 0,
            day: vec![-1],
        }   
    }
    
    fn next(&mut self, price: i32) -> i32 {
        while price >= self.prices[self.prices.len() - 1] {
            self.prices.pop();
            self.day.pop();
        }
        let res = self.idx - self.day[self.day.len() - 1];
        self.day.push(self.idx);
        self.prices.push(price);
        self.idx += 1;
        res
    }   
}
